### PR TITLE
Support shared headroom pool on top of dynamic buffer calculation - swss part

### PR DESCRIPTION
**Pull request by stephenxs**
Signed-off-by: Stephen Sun <stephens@nvidia.com>

**What I did**

**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

File: cfgmgr/buffermgrdyn.cpp [hunk 1]
```diff
@@ -496,9 +527,14 @@ task_process_status BufferMgrDynamic::allocateProfile(const string &speed, const
             return task_process_status::task_need_retry;
         }
 
+        profile.speed = speed;
+        profile.cable_length = cable;
```
Comment (keboliu) [inline cfgmgr/buffermgrdyn.cpp:531]: cable--->cable_len

File: cfgmgr/buffermgrdyn.cpp [hunk 2]
```diff
@@ -336,9 +336,37 @@ void BufferMgrDynamic::recalculateSharedBufferPool()
...
-                if (pool.total_size == pairs[1])
+                if (poolName == INGRESS_LOSSLESS_PG_POOL_NAME)
+                {
+                    if (pairs.size() > 2)
+                    {
+                        auto &xoffStr = pairs[2];
+                        unsigned long xoffNum = atol(xoffStr.c_str());
+                        if (m_mmuSizeNumber > 0 && m_mmuSizeNumber < xoffNum)
+                        {
+                            SWSS_LOG_ERROR("Buffer pool %s: Invalid xoff %s, exceeding the mmu size %s, ignored xoff but the pool size will be updated",
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:350]: as it was.

File: cfgmgr/buffermgrdyn.cpp [hunk 3]
```diff
@@ -766,6 +802,62 @@ task_process_status BufferMgrDynamic::refreshPriorityGroupsForPort(const string
     return task_process_status::task_success;
 }
 
+void BufferMgrDynamic::refreshSharedHeadroomPool(bool enable_state_updated_by_ratio, bool enable_state_updated_by_size)
+{
+    // lossless profiles need to be refreshed only if system is switched between SHP and non-SHP
+    bool need_refresh_profiles = false;
+    bool shp_enabled_by_size = false;
```
[the remark this answers is not in the document]
Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:809]: there are two ways to enable SHP, one is to configure the size of SHP the other is to configure the over subscribe ratio.
so, `shp_enabled_by_size` means shp is enabled by configuring the size of SHP.

File: cfgmgr/buffermgrdyn.cpp [hunk 4]
```diff
@@ -766,6 +803,62 @@ task_process_status BufferMgrDynamic::refreshPriorityGroupsForPort(const string
...
+
+    auto searchRef = m_bufferProfileLookup.find(INGRESS_LOSSLESS_PG_POOL_NAME);
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:812]: shoud be `m_bufferPoolLookup`

File: tests/test_buffer_dynamic.py [hunk 5]
```diff
@@ -408,3 +408,79 @@ def test_nonDefaultAlpha(self, dvs, testlog):
...
+        ingress_lossless_pool_in_appldb['xoff'] = shp_size
+        self.app_db.wait_for_field_match('BUFFER_POOL_TABLE', 'ingress_lossless_pool', ingress_lossless_pool_in_appldb)
+        ingress_lossless_pool_in_asicdb['SAI_BUFFER_POOL_ATTR_XOFF_SIZE'] = shp_size
+        self.asic_db.wait_for_field_match('ASIC_STATE:SAI_OBJECT_TYPE_BUFFER_POOL', self.ingress_lossless_pool_oid, ingress_lossless_pool_in_asicdb)
+
+        # remove over subscribe ratio configuration
+        default_lossless_buffer_parameter['over_subscribe_ratio'] = '0'
+        self.config_db.update_entry('DEFAULT_LOSSLESS_BUFFER_PARAMETER', 'AZURE', default_lossless_buffer_parameter)
+        # check whether shp size has been removed from both asic db and appl db
+        ingress_lossless_pool_in_appldb.pop('xoff')
```
Comment (stephenxs) [inline tests/test_buffer_dynamic.py:479]: `xoff` == '0'

File: tests/test_buffer_dynamic.py [hunk 6]
```diff
@@ -408,3 +408,79 @@ def test_nonDefaultAlpha(self, dvs, testlog):
...
+        self.app_db.wait_for_field_match('BUFFER_POOL_TABLE', 'ingress_lossless_pool', ingress_lossless_pool_in_appldb)
+        ingress_lossless_pool_in_asicdb.pop('SAI_BUFFER_POOL_ATTR_XOFF_SIZE')
+        self.asic_db.wait_for_field_match('ASIC_STATE:SAI_OBJECT_TYPE_BUFFER_POOL', self.ingress_lossless_pool_oid, ingress_lossless_pool_in_asicdb)
+
+        # remove lossless PG 3-4 on interface
+        self.config_db.delete_entry('BUFFER_PG', 'Ethernet0|3-4')
+        dvs.runcmd('config interface startup Ethernet0')
```
Comment (stephenxs) [inline tests/test_buffer_dynamic.py:486]: config interface shutdown Ethernet0

File: cfgmgr/buffer_pool_mellanox.lua [hunk 7]
```diff
@@ -165,7 +196,14 @@ for i = 1, #epools, 1 do
     end
 end
 
+if shp_enabled and shp_size == 0 then
+    accumulative_xoff = math.ceil(accumulative_xoff / over_subscribe_ratio)
```
Comment (stephenxs) [inline cfgmgr/buffer_pool_mellanox.lua:200]: use `shp_size` instead?

File: cfgmgr/buffer_pool_mellanox.lua [hunk 8]
```diff
@@ -165,7 +196,14 @@ for i = 1, #epools, 1 do
...
+else
+    accumulative_xoff = shp_size
+end
+
 local pool_size
+accumulative_occupied_buffer = accumulative_occupied_buffer + accumulative_xoff
```
Comment (stephenxs) [inline cfgmgr/buffer_pool_mellanox.lua:206]: 3 cases:
- SHP enabled by over subscribe ratio, `accumulative_xoff` represents SHP size
- SHP enabled by size, `accumulative_xoff` represents SHP size
- SHP disabled, `accumulative_xoff` represents the accumulative xoff

**Review by keboliu — APPROVED** [review]

File: cfgmgr/buffermgrdyn.cpp [hunk 9]
```diff
@@ -435,14 +466,14 @@ void BufferMgrDynamic::updateBufferProfileToDb(const string &name, const buffer_
...
-        fvVector.emplace_back(make_pair("xon_offset", profile.xon_offset));
+        fvVector.emplace_back("xon_offset", profile.xon_offset);
     }
-    fvVector.emplace_back(make_pair("xoff", profile.xoff));
-    fvVector.emplace_back(make_pair("size", profile.size));
-    fvVector.emplace_back(make_pair("pool", "[" + pg_pool_reference + "]"));
-    fvVector.emplace_back(make_pair(mode, profile.threshold));
+    fvVector.emplace_back("xoff", profile.xoff);
+    fvVector.emplace_back("size", profile.size);
+    fvVector.emplace_back("pool", "[" + pg_pool_reference + "]");
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:475]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 10]
```diff
@@ -476,7 +507,7 @@ void BufferMgrDynamic::updateBufferPgToDb(const string &key, const string &profi
 }
 
 // We have to check the headroom ahead of applying them
-task_process_status BufferMgrDynamic::allocateProfile(const string &speed, const string &cable, const string &mtu, const string &threshold, const string &gearbox_model, string &profile_name)
+task_process_status BufferMgrDynamic::allocateProfile(const string &speed, const string &cable_len, const string &mtu, const string &threshold, const string &gearbox_model, string &profile_name)
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:510]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 11]
```diff
@@ -721,18 +756,20 @@ task_process_status BufferMgrDynamic::refreshPriorityGroupsForPort(const string
...
             // Add the old profile to "to be removed" set
             if (!oldProfile.empty())
+            {
                 profilesToBeReleased.insert(oldProfile);
+                m_bufferProfileLookup[oldProfile].port_pgs.erase(key);
+            }
         }
 
+        // buffer pg needs to be updated as well
+        portPg.running_profile_name = newProfile;
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:771]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 12]
```diff
@@ -811,12 +896,6 @@ task_process_status BufferMgrDynamic::doUpdatePgTask(const string &pg_key, const
         return task_process_status::task_success;
     }
 
-    if (bufferPg.static_configured && bufferPg.dynamic_calculated)
-    {
-        auto &profile = m_bufferProfileLookup[bufferPg.configured_profile_name];
-        profile.port_pgs.insert(pg_key);
-    }
-
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:819]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 13]
```diff
@@ -843,12 +922,6 @@ task_process_status BufferMgrDynamic::doRemovePgTask(const string &pg_key, const
     SWSS_LOG_NOTICE("try removing the original profile %s", bufferPg.running_profile_name.c_str());
     releaseProfile(bufferPg.running_profile_name);
 
-    if (bufferPg.static_configured && bufferPg.dynamic_calculated)
-    {
-        auto &profile = m_bufferProfileLookup[bufferPg.configured_profile_name];
-        profile.port_pgs.erase(pg_key);
-    }
-
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:851]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 14]
```diff
@@ -1099,15 +1192,18 @@ task_process_status BufferMgrDynamic::handlePortTable(KeyOpFieldsValuesTuple &tu
         string &mtu = portInfo.mtu;
         string &speed = portInfo.speed;
 
-        if (cable_length.empty() || speed.empty())
-        {
-            SWSS_LOG_WARN("Cable length for %s hasn't been configured yet, unable to calculate headroom", port.c_str());
-            // We don't retry here because it doesn't make sense until the cable length is configured.
-            return task_process_status::task_success;
-        }
-
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1108]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 15]
```diff
@@ -1099,15 +1192,18 @@ task_process_status BufferMgrDynamic::handlePortTable(KeyOpFieldsValuesTuple &tu
...
+            if (cable_length.empty() || speed.empty())
+            {
+                // we still need to update pool size when port with headroom override is shut down
+                // even if its cable length or speed isn't configured
+                // so cable length and speed isn't tested for shutdown
+                SWSS_LOG_WARN("Cable length for %s hasn't been configured yet, unable to calculate headroom", port.c_str());
+                // We don't retry here because it doesn't make sense until the cable length is configured.
+                return task_process_status::task_success;
+            }
+
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1206]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 16]
```diff
@@ -1295,7 +1413,7 @@ task_process_status BufferMgrDynamic::handleBufferProfileTable(KeyOpFieldsValues
                     profileApp.ingress = true;
                 }
             }
-            fvVector.emplace_back(FieldValueTuple(field, value));
+            fvVector.emplace_back(field, value);
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1416]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 17]
```diff
@@ -1353,13 +1471,20 @@ task_process_status BufferMgrDynamic::handleBufferProfileTable(KeyOpFieldsValues
...
                 }
             }
-        }
 
-        m_applBufferProfileTable.del(profileName);
-        m_stateBufferProfileTable.del(profileName);
+            if (!profile.static_configured || !profile.dynamic_calculated)
+            {
+                m_applBufferProfileTable.del(profileName);
+                m_stateBufferProfileTable.del(profileName);
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1478]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 18]
```diff
@@ -1353,13 +1471,20 @@ task_process_status BufferMgrDynamic::handleBufferProfileTable(KeyOpFieldsValues
...
+            }
 
-        m_bufferProfileLookup.erase(profileName);
-        m_bufferProfileIgnored.erase(profileName);
+            m_bufferProfileLookup.erase(profileName);
+            m_bufferProfileIgnored.erase(profileName);
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Profile %s not found while being removed", profileName.c_str());
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1486]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 19]
```diff
@@ -1418,7 +1549,7 @@ task_process_status BufferMgrDynamic::handleOneBufferPgEntry(const string &key,
                     }
                     else
                     {
-                        // In this case, we shouldn't set the dynamc calculated flat to true
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1421]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 20]
```diff
@@ -1435,7 +1566,7 @@ task_process_status BufferMgrDynamic::handleOneBufferPgEntry(const string &key,
                 bufferPg.static_configured = true;
                 bufferPg.configured_profile_name = profileName;
             }
-            fvVector.emplace_back(FieldValueTuple(field, value));
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1438]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 21]
```diff
@@ -1449,6 +1580,11 @@ task_process_status BufferMgrDynamic::handleOneBufferPgEntry(const string &key,
         if (!ignored && bufferPg.lossless)
         {
             doUpdatePgTask(key, port);
+
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1583]: Enhance

File: cfgmgr/buffermgrdyn.cpp [hunk 22]
```diff
@@ -1589,7 +1729,7 @@ task_process_status BufferMgrDynamic::doBufferTableTask(KeyOpFieldsValuesTuple &
                 transformReference(fvValue(i));
             if (fvField(i) == "profile_list")
                 transformReference(fvValue(i));
-            fvVector.emplace_back(FieldValueTuple(fvField(i), fvValue(i)));
```
Comment (stephenxs) [inline cfgmgr/buffermgrdyn.cpp:1592]: Enhance

**Pull request closed without merging**